File: JPA/sample-projects/shop-api/src/main/java/me/kzv/shopapi/domain/Address.java
package me.kzv.shopapi.domain;

import lombok.Getter;

import javax.persistence.Embeddable;

@Getter
@Embeddable
public class Address {

    private String city;
    private String street;
    private String zipcode;

    protected Address(){} // JPA 는 protected 기본 생성자 까지 허용함.

    public Address(String city, String street, String zipcode) {
        this.city = city;
        this.street = street;
        this.zipcode = zipcode;
    }

    /*
        JPA 는 엔티티나 임베디드 타입을 사용할 때 자바 기본 생성자를 public 또는 protected 로 설정해야 한다.
        이러한 제약을 두는 이유는 JPA 구현 라이브러리가 객체를 생성할 때 자바 리플랙션과 같은 기술을 사용할 수 있도록 지원해야 하기 때문이다.

        자바 리플렉션이란?
        구체적인 클래스 타입을 알지 못해도, 그 클래스의 메소드, 타입, 변수들에 접근할 수 있도록 해주는 자바 API 이다.

        자바 리플렉션 -> 자세한 내용은 https://1-7171771.tistory.com/123 참고
     */
}

File: JPA/kotlin-jpa-querydsl/src/main/kotlin/me/kzv/kotlinjpaquerydsl/entity/Address.kt
package me.kzv.kotlinjpaquerydsl.entity

import jakarta.persistence.Column
import jakarta.persistence.Embeddable

/**
 * https://www.inflearn.com/questions/78106/embeddable-%EC%82%AC%EC%9A%A9%EC%8B%9C-%EC%A7%88%EB%AC%B8-%EC%9E%85%EB%8B%88%EB%8B%A4
 *
 * JPA가 FIELD, 프로퍼티(getter,setter) 엑세스 두가지 방법이 있는데요.
 * 아마 getter 같은게 보여서 정확히 판단이 어려워서 그렇게 나오는 것 같아요.
 * 참고로 프로퍼티 접근은 최근에는 권장하지 않습니다.
 * 해결방법은 2가지가 있습니다.
 * 1. @Access로 필드 접근을 명시한다.
 * 2. 첫 컬럼에 @Column으로 필드 접근 방식을 사용하고 있다는 것을 명시한다.
 *
 */

@Embeddable // 내장 타입
class Address(
    @Column val city: String,
    @Column val street: String,
    @Column val zipcode: String,
)
